aten/src/ATen/native/Histogram.cpp

```diff
@@ -0,0 +1,211 @@
+    TORCH_CHECK(input.dtype() == bins.dtype(), "torch.histogram: input tensor and bins tensor should",
+            " have the same dtype, but got input ", input.dtype(), " and bins ", bins.dtype());
+
+    TORCH_CHECK(bins.dim() == 1, "torch.histogram: bins tensor should have dimension 1,",
+            " but got ", bins.dim(), " dimension");
+
+    TORCH_CHECK(bins.numel() > 0, "torch.histogram: bins tensor should have at least 1 element,",
+            " but got ", bins.numel(), " elements");
+
+    if (weight.has_value()) {
+        TORCH_CHECK(input.dtype() == weight.value().dtype(), "torch.histogram: if weight tensor is provided,"
+                " input tensor and weight tensor should have the same dtype, but got input(", input.dtype(), ")",
+                ", and weight(", weight.value().dtype(), ")");
+
+        TORCH_CHECK(input.sizes() == weight.value().sizes(), "torch.histogram: if weight tensor is provided,"
+                " input tensor and weight tensor should have the same shape, but got input(", input.sizes(), ")",
+                ", and weight(", weight.value().sizes(), ")");
+    }
+}
+
+/* Checks properties of output tensors hist and bin_edges, then resizes them.
+ */
+void histogram_prepare_out(const Tensor& input, int64_t bin_ct,
+        const Tensor& hist, const Tensor& bin_edges) {
+    TORCH_CHECK(input.dtype() == hist.dtype(), "torch.histogram: input tensor and hist tensor should",
+            " have the same dtype, but got input ", input.dtype(), " and hist ", hist.dtype());
+
+    TORCH_CHECK(input.dtype() == bin_edges.dtype(), "torch.histogram: input tensor and bin_edges tensor should",
+            " have the same dtype, but got input ", input.dtype(), " and bin_edges ", bin_edges.dtype());
+
+    TORCH_CHECK(bin_ct > 0,
+            "torch.histogram(): bins must be > 0, but got ", bin_ct);
+
+    at::native::resize_output(hist, bin_ct);
+
+    at::native::resize_output(bin_edges, bin_ct + 1);
+
+    TORCH_CHECK(hist.is_contiguous(), "torch.histogram: hist tensor must be contiguous");
+}
+
+/* Determines the outermost bin edges.
+ */
+std::pair<double, double> select_outer_bin_edges(const Tensor& input,
+        const c10::optional<Scalar>& min, const c10::optional<Scalar>& max) {
+    if (min.has_value() && max.has_value()) {
+        return std::make_pair(min.value().to<double>(), max.value().to<double>());
+    }
+
+    // Default range for empty input matching numpy.histogram's default
+    if (input.numel() == 0) {
+        return std::make_pair(0., 1.);
+    }
+
+    auto extrema = _aminmax(input);
+
+    double leftmost_edge  = min.has_value() ? min.value().to<double>() : std::get<0>(extrema).item<double>();
+    double rightmost_edge = max.has_value() ? max.value().to<double>() : std::get<1>(extrema).item<double>();
+
+    TORCH_CHECK(leftmost_edge <= rightmost_edge, "torch.histogram: min should not exceed max, but got",
```

Comment:
I'm a little confused by this check because it happens only if `aminmax` is called -- but that means that the user can specify min=10 and max=5? Also, isn't this invariant guaranteed by `aminmax`?

Reply:
I believe since we've chosen to take min and max as separate arguments and not as a range tuple, this check can be relevant if the user decides to define one but not the other.

The short-circuit in the case that both min and max are provided is failing to perform this check, though.

Reply:
On further thought things get messy here if we allow, for example, specification of min but not max along with an empty input (cannot deduce max from the input values).  

Numpy doesn't encounter such a case because it accepts a range tuple containing both min and max. [torch.histc](https://github.com/search?p=1&q=torch.histc&type=Code) users generally pass both min and max or neither.

Could we just expect that both min and max are specified if either is, and throw an error otherwise?

Reply:
Requiring both be specified, or switching to range, makes sense to me. Maybe when both are required range is a more natural UX?